models/Image.js

```diff
@@ -1,76 +1,153 @@
+  })
+
+  return Promise.all(promises);
+}
+
+/**
+ *  Build resize metadata.
+ *
+ *  @param {string} filename - uploaded file name
+ *  @param {number} originWidth - image width
+ *  @param {number} originHeight - image height
+ *  @returns {Object}
+ */
+function getResizeMetadata(filename, originWidth, originHeight) {
+  const [filenameWithoutExt, ext] = filename.split('.');
+  const targets = {};
+
+  resizedOpts.forEach(resizedOpt => {
+    let [resolution, resizeWidth] = resizedOpt;
+    let resizeHeight = 0;
+
+    if (resizeWidth < originWidth) {
+      resizeHeight = Math.round((resizeWidth / originWidth) * originHeight);
+    } else {
+      resizeWidth = originWidth;
+      resizeHeight = originHeight;
+    }
+
+    targets[resolution] = {
+      url: `${resizeImageURLPrefix}${filenameWithoutExt}-${resolution}.${ext}`,
+      width: resizeWidth,
+      height: resizeHeight,
+    };
+  });
+
+  return targets;
+}
 
 Image.add({
-    image: {
-        type: Types.GcsImage,
-        initial: true,
-        autoCleanup: true,
-        datePrefix: 'YYYYMMDDHHmmss',
-        // TODO move these settings to config
-        bucket: bucket,
-        destination: 'images/',
-        publicRead: true,
-        resize: resizeImage,
-        resizeOpts: [{
-            target: 'desktop',
-            width: 2000,
-            height: null,
-            options: {}
-        }, {
-            target: 'tablet',
-            width: 1200,
-            height: null,
-            options: {}
-        }, {
-            target: 'mobile',
-            width: 800,
-            height: null,
-            options: {}
-        }, {
-            target: 'w400',
-            width: 400,
-            height: null,
-            options: {}
-        }, {
-            target: 'tiny',
-            width: 150,
-            height: null,
-            options: {}
-        }],
-        extractIPTC: extractIPTC
-    },
-    description: {
-        type: String,
-        index: true
-    },
-    copyright: {
-        type: Types.Select,
-        options: 'Creative-Commons, Copyrighted',
-        default: 'Copyrighted', index: true
-    },
-    keywords: {
-        type: String,
-        index: true
-    },
-    sale: {
-        type: Boolean,
-        index: true
+  image: {
+    type: Types.GcsImage,
+    initial: true,
+    autoCleanup: true,
+    datePrefix: 'YYYYMMDDHHmmss',
+    bucket: bucket,
+    destination: '/',
+    publicRead: false,
+    hooks: {
+      postUpload: async (imageData) => {
+        if (typeof imageData !== 'object' || imageData === null) {
+          return imageData;
+        }
+
+        const { filepath, filename, height, width } = imageData;
+
+        if (typeof filename !== 'string' || filename === '') {
+          return imageData;
+        }
+
+        try {
+          await publishResizeTasks(filename);
+        } catch (err) {
+          throw errors.helpers.wrap(err, 'Error', 'Received error while publishing resize events.')
```

Comment:
Q: How does the error be handled in Keystone client?

Reply:
https://github.com/twreporter/keystone/pull/154/files#diff-ae08a0f006df447dffc0742f2140bfaf13837aeef8ec2170dc4880c2b93f101dR306-R315
Keystone will delete that uploaded file from GCS, and tell end user the uploading fails.